### PR TITLE
bpo-28254: Add a C-API for controlling the state of the garbage collector

### PR DESCRIPTION
**Pull request by scoder**
https://bugs.python.org/issue28254

### CODE REVIEW

**Comment by pablogsal** [comment]
Looks good, I will review it in more detail today, but please, add an entry to the what's new document for 3.10 in the meantime.

File: Misc/NEWS.d/next/C API/2021-04-28-12-33-44.bpo-28254.a2561e.rst
Comment (pablogsal) [inline Misc/NEWS.d/next/C API/2021-04-28-12-33-44.bpo-28254.a2561e.rst:1]: Please, use markup here. Example:
```

:c:func:`my_func`
```

Comment (scoder) [reply Misc/NEWS.d/next/C API/2021-04-28-12-33-44.bpo-28254.a2561e.rst:1]: Done.

File: Doc/whatsnew/3.10.rst
Comment (pablogsal) [inline Doc/whatsnew/3.10.rst]:
```suggestion
  These functions allow to activate, deactivate and query the state of the garbage collector from C code without
  having to import the :mod:`gc` module.
```

File: Modules/_testcapimodule.c
Comment (pablogsal) [inline Modules/_testcapimodule.c]:
```suggestion
    if (!PyGC_IsEnabled()) {
        goto failed;
    }
```

File: Modules/_testcapimodule.c
Comment (pablogsal) [inline Modules/_testcapimodule.c]:
```suggestion
    if (PyGC_IsEnabled()) {
        goto failed;
    }
```

File: Modules/_testcapimodule.c
Comment (pablogsal) [inline Modules/_testcapimodule.c]:
```suggestion
    if (!PyGC_IsEnabled()) {
        goto failed;
    }
```

**Review by pablogsal — APPROVED** [review]

**Comment by pablogsal** [comment]
Thanks for the PR @scoder! Make sure to merge it before end of the week 😉

**Comment by scoder** [comment]
I changed the functions to return the previous state. That's a common idiom in C, and I think a useful one in this case.

File: Doc/whatsnew/3.10.rst
Comment (vstinner) [inline Doc/whatsnew/3.10.rst:1697]: Please document these functions but also PyGC_Collect().

https://docs.python.org/dev/c-api/gcsupport.html may be good home for this.

I'm disappointed, PyGC_Collect() is not documented at all!

File: Modules/_testcapimodule.c
Comment (vstinner) [inline Modules/_testcapimodule.c:155]: Bugs are unlikely. IMO using assert() is good enough for _testcapi tests. You can leave the code as it is ;-)

See for example test_refcount_funcs() which uses assert().

File: Doc/whatsnew/3.10.rst
Comment (scoder) [reply Doc/whatsnew/3.10.rst:1697]: Yeah, I was kind of considering that they might have been intentionally undocumented. But probably not. I'll write something up.

File: Modules/_testcapimodule.c
Comment (scoder) [reply Modules/_testcapimodule.c:155]: I was thinking that assertions can be compiled out, which is decidedly not intended here.

**Comment by scoder** [comment]
Are the C functions prototypes defined in the right place? Limited API, header file hierarchy, etc.?

File: Modules/_testcapimodule.c
Comment (vstinner) [reply Modules/_testcapimodule.c:155]: In general yes, but _testcapimodule.c starts with:

```
/* Always enable assertions */
#undef NDEBUG
```

**Comment by vstinner** [comment]
> Are the C functions prototypes defined in the right place? Limited API, header file hierarchy, etc.?

You added them to the stable ABI and so you should run "regen-limited-abi".

File: Doc/c-api/gcsupport.rst
Comment (vstinner) [inline Doc/c-api/gcsupport.rst]: Errors cannot be ignored and printed at the same time :-) Exceptions are logged with sys.unraisablehook, you can add a link using:

```
:data:`sys.unraisablehook`
```

File: Doc/c-api/gcsupport.rst
Comment (vstinner) [inline Doc/c-api/gcsupport.rst:194]: @pablogsal: Do we want to document that it's safe to call this function with an exception raised, and that the exception is saved and then restored?

File: Doc/c-api/gcsupport.rst
Comment (vstinner) [inline Doc/c-api/gcsupport.rst]: Hum, uncollectable objects are stored in gc.garbage no? Shoud we document that?

The C code uses "unreachable objects that couldn't be collected" in a comment instead of "uncollectable objects". Maybe it's a more explicit definition.

**Comment by scoder** [comment]
> > Are the C functions prototypes defined in the right place? Limited API, header file hierarchy, etc.?
> 
> You added them to the stable ABI and so you should run "regen-limited-abi".

That added more than the three. I only committed the ones I added here in order to keep the PR clean.
I'll check where the others came from.

File: Doc/whatsnew/3.10.rst
Comment (scoder) [reply Doc/whatsnew/3.10.rst:1697]: Done.

File: Doc/c-api/gcsupport.rst
Comment (vstinner) [inline Doc/c-api/gcsupport.rst]:
```suggestion
   Performs a full garbage collection, if the garbage collector is enabled: same as calling func:`gc.collect`.
```

You may mention that free lists of built-in types are also cleared.

File: Modules/_testcapimodule.c
Comment (scoder) [reply Modules/_testcapimodule.c:155]: Good to know! I'd keep the test as it is though. It doesn't really hurt, I think.

File: Doc/c-api/gcsupport.rst
Comment (vstinner) [inline Doc/c-api/gcsupport.rst]:
```suggestion

   If the garbage collector is disabled or already collecting,
   does nothing and returns ``0`` immediately.
   
```

File: Doc/c-api/gcsupport.rst
Comment (vstinner) [inline Doc/c-api/gcsupport.rst:183]: Once you writes "Performs" and once "Enable" (no S). I suggest to avoid the S, most documentations avoid it.

**Comment by vstinner** [comment]
> That added more than the three. I only committed the ones I added here in order to keep the PR clean.
I'll check where the others came from.

Right, thanks: https://bugs.python.org/issue43795#msg392212

File: Doc/c-api/gcsupport.rst
Comment (scoder) [reply Doc/c-api/gcsupport.rst]: Done.

File: Doc/c-api/gcsupport.rst
Comment (scoder) [reply Doc/c-api/gcsupport.rst]: I don't think this C-API function is the right place to document these details. That's what the `gc` module is for.

File: Doc/c-api/gcsupport.rst
Comment (scoder) [reply Doc/c-api/gcsupport.rst]: It's actually not the same as `gc.collect()`. The latter doesn't care about `gc.enabled`, whereas `PyGC_Collect()` honours it.

Comment (vstinner) [reply Doc/c-api/gcsupport.rst]: > I don't think this C-API function is the right place to document these details. That's what the gc module is for.

I agree that gc.collect() documentation must be the most complete. But in this case, mention that PyGC_Collect is similar to gc.collect() but just document changes between PyGC_Collect() and gc.collect(). As you wrote, the main one is that PyGC_Collect() honors gc.enabled.

File: Doc/c-api/gcsupport.rst
Comment (vstinner) [inline Doc/c-api/gcsupport.rst]:
```suggestion
   Enable the garbage collector: similar to :func:`gc.enable`.
```

File: Doc/c-api/gcsupport.rst
Comment (vstinner) [inline Doc/c-api/gcsupport.rst]:
```suggestion
   Disable the garbage collector: similar to :func:`gc.disable`.
```

File: Doc/c-api/gcsupport.rst
Comment (vstinner) [inline Doc/c-api/gcsupport.rst]:
```suggestion
   Query the state of the garbage collector: similar to :func:`gc.isenabled`.
```

File: Doc/c-api/gcsupport.rst
Comment (pablogsal) [reply Doc/c-api/gcsupport.rst]: I can document `PyGC_Collect` later, let's focus this PR on the new APIs.

Comment (vstinner) [reply Doc/c-api/gcsupport.rst]: I'm fine with documenting this "complex" function in a separated PR.

File: Doc/c-api/gcsupport.rst
Comment (vstinner) [inline Doc/c-api/gcsupport.rst]: Oh, the markup should be at the same indentation level than the paragraph. Also "versionadded" is better than "versionchanged".

```suggestion
   Returns the previous state, 0 for disabled and 1 for enabled.

   .. versionadded:: 3.10
```

Same remark for the 2 other added functions.

**Review by vstinner — APPROVED** [review]
LGTM but please fix the Sphinx markup.

File: Doc/c-api/gcsupport.rst
Comment (scoder) [reply Doc/c-api/gcsupport.rst]: I added a sentence to compare the two. Feel free to take another pass later.

File: Doc/c-api/gcsupport.rst
Comment (scoder) [reply Doc/c-api/gcsupport.rst]: Right. I had thought of it and then still ended up copynpasting… :-/

File: Doc/c-api/gcsupport.rst
Comment (vstinner) [inline Doc/c-api/gcsupport.rst]: Please fix also the indentation.

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
Merged, thanks @scoder!

I'm not really excited by adding a C API for each Python function, but this one seems common enough to justify adding an API. Also, @pablogsal liked the idea ;-)

**Comment by pablogsal** [comment]
> Merged, thanks @scoder!
> 
> I'm not really excited by adding a C API for each Python function, but this one seems common enough to justify adding an API. Also, @pablogsal liked the idea ;-)

The GC gods we all serve are pleased with the new API ;)

**Comment by scoder** [comment]
Thanks Victor and Pablo!

**Comment by vstinner** [comment]
I created PR #25693 which shows how useful are these new C API functions ;-) Ok, maybe, it makes the C code a little bit simpler ;-)

**Comment by scoder** [comment]
35 lines added, 116 lines saved. Seems a good deal.

**Comment by markshannon** [comment]
FYI, this has broken CI 
Example: https://github.com/python/cpython/pull/25719/checks?check_run_id=2466874649
Proposed fix:
https://github.com/python/cpython/pull/25716

**Comment by encukou** [comment]
Yest, this adds things to `Doc/data/stable_abi.dat` but not ` PC/python3dll.c`, which is quite wrong.
Meanwhile, I merged the part of PEP-652 that will prevent these from going out of sync (by generating them both from a single manifest), but the CI on this PR ran before that and the merge happened after.

From now on, CI will be fail on the pull request when issues like this happen, but this one's timing was unfortunate.

**Comment by pablogsal** [comment]
Ok, then let's merge https://github.com/python/cpython/pull/25720 to unblock this

**Pull request merged**
